### PR TITLE
[local_auth] updated getAvailableBiometrics behavior on Android

### PR DESCRIPTION
**Pull request by RafaRuiz**
**[Continuing PR #4874 from @darwin-morocho]**

On Android `canCheckBiometrics` returns true if the device doesn't have any biometric enrolled (For more info check https://github.com/flutter/flutter/issues/45497#issuecomment-1007310214). Now with this PR `canCheckBiometrics` returns false on Android if device has the biometric hardware but doesn't have any biometric enrolled.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

**Comment by RafaRuiz** [comment]
@stuartmorgan  any thoughts?

File: packages/local_auth/local_auth/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
-## NEXT
+## 1.1.12
 
 * Adds OS version support information to README.
+* Fixes `getAvailableBiometrics` behavior on Android. Now it returns an empty
+list when device has the biometric hardware but doesn't have any biometric
+enrolled. For more information check [this](https://github.com/flutter/flutter/issues/45497#issuecomment-1007310214).
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/CHANGELOG.md:6]: The last sentence can be removed; that comment doesn't really have any information that the CHANGELOG entry doesn't.

File: packages/local_auth/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java [hunk 2]
```diff
@@ -269,13 +270,18 @@ private void getAvailableBiometrics(final Result result) {
     if (activity == null || activity.isFinishing()) {
       return biometrics;
     }
+
+    if (!canAuthenticateWithBiometrics()) {
+      return biometrics;
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:275]: You haven't addressed my feedback from the previous PR, so this has the same bug as earlier versions of that PR. Please read my comments there regarding the desired behavior with respect to empty lists vs `["undefined"]`.

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 3]
```diff
@@ -20,45 +23,83 @@
...
-    plugin.onMethodCall(new MethodCall("isDeviceSupported", null), mockResult);
-    verify(mockResult).success(false);
-  }
 
-  @Test
-  public void onDetachedFromActivity_ShouldReleaseActivity() {
-    final Activity mockActivity = mock(Activity.class);
-    final ActivityPluginBinding mockActivityBinding = mock(ActivityPluginBinding.class);
-    when(mockActivityBinding.getActivity()).thenReturn(mockActivity);
+  private LocalAuthPlugin plugin = new LocalAuthPlugin();
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:33]: The object under test should not be part of the test fixture; that's an anti-pattern in unit testing.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 4]
```diff
@@ -20,45 +23,83 @@
...
     plugin.onAttachedToEngine(mockPluginBinding);
     plugin.onAttachedToActivity(mockActivityBinding);
     assertNotNull(plugin.getActivity());
 
     plugin.onDetachedFromActivity();
     assertNull(plugin.getActivity());
   }
+
+  @Test
+  public void getAvailableBiometrics_withHardwareButNothingEnrolled_ShouldReturnEmpty() {
```
[the remark this answers is not in the document]
Comment (RafaRuiz) [reply packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:85]: What I read here is:
- Given that I have hardware available for biometrics.
- When I haven't enrolled to any.
- Then result of the available biometrics is empty.

What am I missing here?

[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java]: See https://github.com/flutter/plugins/pull/4874#discussion_r814094211.

Alternately, now that this package has been federated and switched to an in-package platform channel you could rework the interface such that this was correct without regression other plugins, but this PR currently isn't doing that either.

**Comment by RafaRuiz** [comment]
In my latest commit, I'm reverting so only the `local_auth_android` is changed to version 1.0.1,
then another PR will proceed after that by changing `local_auth` when `local_auth_android` is published to 1.0.1

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (RafaRuiz) [reply packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java]: Ok, please check the latest version :)
I apologise if this is not very clear as I'm adopting a PR from 1.x to 2.x and it's a bit of a mess :D

File: packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java [hunk 5]
```diff
@@ -269,13 +270,23 @@ private ArrayList<String> getAvailableBiometrics() {
     if (activity == null || activity.isFinishing()) {
       return biometrics;
     }
+
+    if (!hasBiometricHardware()) {
+      biometrics.add("undefined");
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:275]: This is the opposite of what the Dart code is expecting, so this doesn't fix my comment, and also breaks the reverse case.

**Review by stuartmorgan — COMMENTED** [review]
It seems like there's still a fundamental miscommunication about how this interacts with the Dart side of the plugin.

Given that https://github.com/flutter/plugins/pull/5309 was just opened with correct behavior and test structure, I'm going to close in favor of that PR, but thanks for the submission!

**Pull request closed without merging**